File: internal/shellwire/shellwire_test.go
package shellwire_test

import (
	"testing"

	"github.com/icholy/xagent/internal/shellwire"
	"gotest.tools/v3/assert"
)

func TestDataRoundTrip(t *testing.T) {
	t.Parallel()
	// Arrange: arbitrary non-UTF-8 bytes, since PTY output is not text-safe.
	payload := []byte{0x00, 0x01, 0xff, 0xfe, 0x80, 'h', 'i', 0x00}

	// Act
	frame, err := shellwire.Parse(shellwire.Data(payload))

	// Assert
	assert.NilError(t, err)
	assert.Equal(t, frame.Type, shellwire.TypeData)
	assert.DeepEqual(t, frame.Payload, payload)
}

func TestDataEmptyRoundTrip(t *testing.T) {
	t.Parallel()
	// Act
	frame, err := shellwire.Parse(shellwire.Data(nil))

	// Assert
	assert.NilError(t, err)
	assert.Equal(t, frame.Type, shellwire.TypeData)
	assert.Equal(t, len(frame.Payload), 0)
}

func TestResizeRoundTrip(t *testing.T) {
	t.Parallel()
	// Arrange
	tests := []struct {
		name       string
		rows, cols uint16
	}{
		{"typical", 24, 80},
		{"zero", 0, 0},
		{"max", 65535, 65535},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Act
			frame, err := shellwire.Parse(shellwire.Resize(tt.rows, tt.cols))
			assert.NilError(t, err)
			assert.Equal(t, frame.Type, shellwire.TypeResize)
			rows, cols, err := frame.ResizeDims()

			// Assert
			assert.NilError(t, err)
			assert.Equal(t, rows, tt.rows)
			assert.Equal(t, cols, tt.cols)
		})
	}
}

func TestExitRoundTrip(t *testing.T) {
	t.Parallel()
	// Arrange
	for _, code := range []int{0, 1, 42, 130, 255} {
		// Act
		frame, err := shellwire.Parse(shellwire.Exit(code))
		assert.NilError(t, err)
		assert.Equal(t, frame.Type, shellwire.TypeExit)
		got, err := frame.ExitCode()

		// Assert
		assert.NilError(t, err)
		assert.Equal(t, got, code)
	}
}

func TestPingRoundTrip(t *testing.T) {
	t.Parallel()
	// Act
	frame, err := shellwire.Parse(shellwire.Ping())

	// Assert
	assert.NilError(t, err)
	assert.Equal(t, frame.Type, shellwire.TypePing)
	assert.Equal(t, len(frame.Payload), 0)
}

func TestParseRejectsEmpty(t *testing.T) {
	t.Parallel()
	// Act
	_, err := shellwire.Parse(nil)

	// Assert
	assert.ErrorContains(t, err, "empty frame")
}

func TestResizeDimsRejectsWrongType(t *testing.T) {
	t.Parallel()
	// Arrange
	frame, err := shellwire.Parse(shellwire.Data([]byte("x")))
	assert.NilError(t, err)

	// Act
	_, _, err = frame.ResizeDims()

	// Assert
	assert.ErrorContains(t, err, "not a resize frame")
}

func TestExitCodeRejectsWrongType(t *testing.T) {
	t.Parallel()
	// Arrange
	frame, err := shellwire.Parse(shellwire.Data([]byte("x")))
	assert.NilError(t, err)

	// Act
	_, err = frame.ExitCode()

	// Assert
	assert.ErrorContains(t, err, "not an exit frame")
}

File: webui/src/lib/shellwire.test.ts
import { describe, it, expect } from 'vitest'
import { FrameType, encodeData, encodeResize, parse } from './shellwire'

describe('shellwire', () => {
  it('encodes a data frame as [0x00, ...bytes]', () => {
    const payload = new Uint8Array([0x00, 0xff, 0x80, 'h'.charCodeAt(0)])
    const frame = encodeData(payload)
    expect(frame[0]).toBe(FrameType.Data)
    expect(Array.from(frame.subarray(1))).toEqual(Array.from(payload))
  })

  it('round-trips a data frame with arbitrary non-UTF-8 bytes', () => {
    const payload = new Uint8Array([0x00, 0x01, 0xff, 0xfe, 0x80])
    const parsed = parse(encodeData(payload))
    expect(parsed.type).toBe(FrameType.Data)
    expect(Array.from(parsed.data!)).toEqual(Array.from(payload))
  })

  it('encodes resize as big-endian rows then cols', () => {
    const frame = encodeResize(24, 80)
    expect(Array.from(frame)).toEqual([FrameType.Resize, 0x00, 0x18, 0x00, 0x50])
  })

  it('round-trips resize dimensions', () => {
    const parsed = parse(encodeResize(50, 132))
    expect(parsed.type).toBe(FrameType.Resize)
    expect(parsed.rows).toBe(50)
    expect(parsed.cols).toBe(132)
  })

  it('decodes an exit frame with a signed code', () => {
    // Go's shellwire.Exit(-1) encodes 0xffffffff; parse must read it as int32.
    const msg = new Uint8Array([FrameType.Exit, 0xff, 0xff, 0xff, 0xff])
    const parsed = parse(msg)
    expect(parsed.type).toBe(FrameType.Exit)
    expect(parsed.code).toBe(-1)
  })

  it('decodes a non-negative exit code', () => {
    const msg = new Uint8Array([FrameType.Exit, 0x00, 0x00, 0x00, 0x7f])
    expect(parse(msg).code).toBe(127)
  })

  it('decodes a ping frame with no payload', () => {
    const parsed = parse(new Uint8Array([FrameType.Ping]))
    expect(parsed.type).toBe(FrameType.Ping)
  })

  it('throws on an empty frame', () => {
    expect(() => parse(new Uint8Array())).toThrow(/empty frame/)
  })

  it('throws on a malformed resize payload', () => {
    expect(() => parse(new Uint8Array([FrameType.Resize, 0x00]))).toThrow(/resize payload/)
  })
})
